### PR TITLE
More async methods added

### PR DESCRIPTION
**Pull request by lcygan-gain**
Work In Progress

### CODE REVIEW

File: src/nms-api/ISession.cs [hunk 1]
```diff
@@ -16,6 +16,7 @@
  */
 
 using System;
+using System.Threading.Tasks;
```
Comment (michaelandrepearce) [inline src/nms-api/ISession.cs:19]: What about CreateMessage do we want async varients on these? I know for AMQP its probably super fast to return as its all done within the process, but in some other protocols hilariously to create a message a remote call is made (yeah! its shocking but the old protocols are old for a reason!)

File: src/nms-api/NMSConnectionFactory.cs [hunk 2]
```diff
@@ -424,6 +425,36 @@ public INMSContext CreateContext(string userName, string password, Acknowledgeme
...
+        {
+            return this.factory.CreateContextAsync(acknowledgementMode);
+        }
+
+        public Task<INMSContext> CreateContextAsync(string userName, string password)
+        {
+            return this.factory.CreateContextAsync(userName, password);
+        }
+
+        public Task<INMSContext> CreateContextAsync(string userName, string password, AcknowledgementMode acknowledgementMode)
```
Comment (michaelandrepearce) [inline src/nms-api/NMSConnectionFactory.cs:453]: What about INMSProducer and INMSConsumer interfaces?

File: src/nms-api/ISession.cs (hunk 1 above)
Comment (michaelandrepearce) [inline src/nms-api/ISession.cs:19]: CreateQueue and CreateTopic (for same reason above)

File: src/nms-api/IConnection.cs [hunk 3]
```diff
@@ -95,16 +96,22 @@ public interface IConnection : IDisposable, IStartable, IStoppable
         /// </summary>
         ISession CreateSession();
 
+        Task<ISession> CreateSessionAsync();
```
Comment (Havret) [inline src/nms-api/IConnection.cs:99]: It would be wise to have a CancellationToken as a parameter to all `*Async` methods.

**Pull request closed without merging**